components/accelerator/nvidia/error/xid/component.go

```diff
@@ -5,150 +5,242 @@ package xid
-		log.Logger.Debugw("nothing found in last state (no data collected yet)", "component", nvidia_component_error_xid_id.Name)
-		return o, nil
-	}
+func (c *XIDComponent) States(_ context.Context) ([]components.State, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return []components.State{c.currState}, nil
+}
 
-	// something went wrong in the poller
-	// just return an error to surface the issue
+func (c *XIDComponent) Events(ctx context.Context, since time.Time) ([]components.Event, error) {
+	var ret []components.Event
+	events, err := c.store.Get(ctx, since)
 	if err != nil {
 		return nil, err
 	}
-
-	ev, ok := last.Output.(*nvidia_query_nvml.XidEvent)
-	if !ok { // shoild never happen
-		return nil, fmt.Errorf("invalid output type: %T, expected nvidia_query_nvml.XidEvent", last.Output)
+	for _, event := range events {
+		ret = append(ret, resolveXIDEvent(event))
 	}
-	if ev != nil && ev.Xid > 0 {
-		o.NVMLXidEvent = ev
+	return ret, nil
+}
 
-		if lerr := c.poller.LastError(); lerr != nil {
-			log.Logger.Warnw("last query failed -- returning cached, possibly stale data", "error", lerr)
-		}
+func (c *XIDComponent) Metrics(ctx context.Context, since time.Time) ([]components.Metric, error) {
+	log.Logger.Debugw("querying metrics", "since", since)
 
-		lastSuccessPollElapsed := time.Now().UTC().Sub(ev.Time.Time)
-		if lastSuccessPollElapsed > 2*c.poller.Config().Interval.Duration {
-			log.Logger.Warnw("last poll is too old", "elapsed", lastSuccessPollElapsed, "interval", c.poller.Config().Interval.Duration)
-		}
-	}
+	return nil, nil
+}
 
-	return o, nil
+func (c *XIDComponent) Close() error {
+	log.Logger.Debugw("closing XIDComponent")
+	// safe to call stop multiple times
+	c.poller.Stop(nvidia_component_error_xid_id.Name)
+	c.cancel()
+	return nil
 }
 
-func (c *component) Events(ctx context.Context, since time.Time) ([]components.Event, error) {
-	o, err := c.tailScan()
-	if err != nil {
-		return nil, err
+func (c *XIDComponent) start(watcher pkg_dmesg.Watcher) {
+	ticker := time.NewTicker(DefaultStateUpdatePeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.rootCtx.Done():
+			return
+		case <-ticker.C:
+			if err := c.updateCurrentState(); err != nil {
+				log.Logger.Errorw("failed to fetch current events", "error", err)
+				continue
+			}
+		case newEvent := <-c.extraEventCh:
+			if newEvent == nil {
+				continue
+			}
+			if err := c.store.Insert(c.rootCtx, *newEvent); err != nil {
+				log.Logger.Errorw("failed to create event", "error", err)
+				continue
+			}
+			events, err := c.store.Get(c.rootCtx, time.Time{})
+			if err != nil {
+				log.Logger.Errorw("failed to get all events", "error", err)
+				continue
+			}
+			c.mu.Lock()
+			c.currState = EvolveHealthyState(events)
+			c.mu.Unlock()
+		case dmesgLine := <-watcher.Watch():
+			log.Logger.Debugw("dmesg line", "line", dmesgLine)
+			ev, err := nvidia_query_xid.ParseDmesgLogLine(metav1.Time{Time: dmesgLine.Timestamp}, dmesgLine.Content)
```

Comment:
We should now use 

https://github.com/leptonai/gpud/blob/02d18149b2f776b9d2f6fc562b2013bca3858556/components/accelerator/nvidia/query/xid/dmesg/dmesg.go#L61-L64